### PR TITLE
Custom Stereo nodelet

### PR DESCRIPTION
**Pull request by servos**
This is still very very rough but it currently get the job done.

This combines the two cameras into a single nodelet and uses software triggering to synchronize them.

@efernandez

### CODE REVIEW

File: pointgrey_camera_driver/nodelet_plugins.xml [hunk 1]
```diff
@@ -4,6 +4,12 @@
   </class>
 </library>
 
+<library path="lib/libPointGreyCustomStereoNodelet">
+  <class name="pointgrey_camera_driver/PointGreyCustomStereoNodelet" type="pointgrey_camera_driver::PointGreyCustomStereoNodelet" base_class_type="nodelet::Nodelet">
+    <description>This is the nodelet for the Point Grey Camera Driver.</description>
```
Comment (efernandez) [inline pointgrey_camera_driver/nodelet_plugins.xml]: pedantic: `This is the nodelet for the Point Grey Custom Stereo Camera Driver.`

File: pointgrey_camera_driver/src/custom_stereo.cpp [hunk 2]
```diff
@@ -0,0 +1,625 @@
...
+              if (!left_cam_.trigger() || !right_cam_.trigger())
+              {
+                NODELET_ERROR("Could not trigger camera");
+              }
+            }
+
+            ros::Time stamp = ros::Time::now();
+            
+            // Get Images
+            if (!publishImage("left", left_cam_, left_cinfo_, left_it_pub_, left_pub_, stamp)
```
Comment (efernandez) [inline pointgrey_camera_driver/src/custom_stereo.cpp:523]: :+1: For using the logic from the monocular cameras, that adds recovery when the cameras fail (stereo didn't have that).

:+1: For adding `publishImage`, we should do the same in the monocular camera (not in this PR)

File: pointgrey_camera_driver/src/custom_stereo.cpp [hunk 3]
```diff
@@ -0,0 +1,625 @@
...
+            ros::Duration(1.0).sleep(); // sleep for one second each time
+          }
+
+          break;
+        case STARTED:
+          {
+            // Trigger cameras if using software triggering
+            if (config_.enable_trigger == true && config_.trigger_source == "software")
+            {
+              if (!left_cam_.trigger() || !right_cam_.trigger())
```
Comment (efernandez) [inline pointgrey_camera_driver/src/custom_stereo.cpp:512]: optional: Maybe split in two, so you can mentioned in the error message which camera (left or right) couldn't be triggered.

File: pointgrey_camera_driver/src/custom_stereo.cpp [hunk 4]
```diff
@@ -0,0 +1,625 @@
...
+              {
+                NODELET_ERROR("Could not trigger camera");
+              }
```
Comment (efernandez) [inline pointgrey_camera_driver/src/custom_stereo.cpp:517]: Does it make sense to continue when one of the cameras isn't triggered?

Maybe just do nothing, or go into a recovery state. You know better how the triggering behaves, so I leave this to you.

**Review by efernandez — APPROVED** [review]

**Comment by efernandez** [comment]
Looks good to me.

Let me know if you'd do any of the minor comments I mentioned or not. Then I'd merge it, or just merge it yourself if you have merge/write access to the repo.

@mikepurvis , do you want to add something?

File: pointgrey_camera_driver/src/custom_stereo.cpp [hunk 5]
```diff
@@ -0,0 +1,625 @@
...
+Neither the name of the Carnegie Mellon University nor the names of its contributors may be used to endorse or promote products derived from this software without specific prior written permission.
+THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+*/
+
+
+
+/**
+   @file nodelet.cpp
+   @author Chad Rockey
+   @date July 13, 2011
```
Comment (mikepurvis) [inline pointgrey_camera_driver/src/custom_stereo.cpp]: Please update the date and add yourself as an author (Chad should probably remain given that this is clearly derived work).

File: pointgrey_camera_driver/src/custom_stereo.cpp [hunk 6]
```diff
@@ -0,0 +1,625 @@
...
+      {
+        binning_x_ = 0;
+        binning_y_ = 2;
+      }
+      else
+      {
+        binning_x_ = 0;
+        binning_y_ = 0;
+      }
+      */
```
Comment (mikepurvis) [inline pointgrey_camera_driver/src/custom_stereo.cpp]: Remove this if you're not using it.

**Comment by servos** [comment]
I am going to decline this PR because it turned out the software sync doesn't actually work that well and this is a bit of a hack job as it is.

**Comment by efernandez** [comment]
OK, but the recovery code should be added to the stereo camera. That would be another PR though, with changes into the stereo driver.

**Comment by servos** [comment]
We don't currently use the stereo driver at all so TBH it is not a high priority.

**Pull request closed without merging**